quickwit-common/src/uri.rs

```diff
@@ -18,46 +18,97 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 use std::env;
+use std::fmt::Display;
 use std::path::{Component, Path, PathBuf};
 
 use anyhow::Context;
 
 /// Default file protocol `file://`
 const FILE_PROTOCOL: &str = "file";
 
-/// Normalizes a URI
-/// A `file://` protocol is assumed if not specified.
-/// File URIs are resolved (normalised) relative to the current working directory
-/// unless an absolute path is specified.
-/// Handles special characters like (~, ., ..)
-pub fn normalize_uri(uri: &str) -> anyhow::Result<String> {
-    let (protocol, mut path) = match uri.split_once("://") {
-        None => (FILE_PROTOCOL, uri.to_string()),
-        Some((protocol, path)) => (protocol, path.to_string()),
-    };
-
-    let current_dir = env::current_dir().context("Could not fetch the current directory.")?;
-
-    if protocol == FILE_PROTOCOL {
-        if path.starts_with('~') {
-            let home_dir_path = home::home_dir()
-                .context("Could not fetch the home directory.")?
+const PROTOCOL_SEPARATOR: &str = "://";
+
+/// Encapsulates the URI type.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Uri {
+    uri: String,
+    protocol_idx: usize,
+}
+
+impl Uri {
+    /// Tries to to construct a Uri from the raw string.
+    /// A `file://` protocol is assumed if not specified.
+    /// File URIs are resolved (normalised) relative to the current working directory
+    /// unless an absolute path is specified.
+    /// Handles special characters like (~, ., ..)
+    pub fn try_new(uri: &str) -> anyhow::Result<Self> {
+        let (protocol, mut path) = match uri.split_once(PROTOCOL_SEPARATOR) {
+            None => (FILE_PROTOCOL, uri.to_string()),
+            Some((protocol, path)) => (protocol, path.to_string()),
+        };
+
+        if protocol == FILE_PROTOCOL {
+            let current_dir =
+                env::current_dir().context("Could not fetch the current directory.")?;
+
+            if path.starts_with('~') {
+                let home_dir_path = home::home_dir()
+                    .context("Could not fetch the home directory.")?
+                    .to_string_lossy()
+                    .to_string();
+
+                let user_name_dir = format!("~{}", whoami::username());
```

Comment:
I think the ~username stuff is still not correct.
Note that we do not need to support it... but I don't want us to do random stuff on it
so your initial code was doing 

if path.startswith("~") {
   // ... logic to do the replace.
}
that would produce very weird and unpredictable result if someone used the syntax above.
~evance/ on my computer would become /home/fulmicotonevance /home/fulmicoton/evance depending on the body.
Your new code only handles ~current_user
We can just keep the previous code, and return an explicit error ("This path syntax: <user_input> is not supported.") if the input starts by ~ and is not followed by a separator.